lib/web_ui/lib/src/engine/bitmap_canvas.dart

```diff
@@ -285,37 +296,107 @@ class BitmapCanvas extends EngineCanvas {
+      drawRect(ui.Rect.fromLTWH(0, 0, _bounds.width, _bounds.height),
+        paintData);
+    } else {
+      _canvasPool.drawColor(color, blendMode);
+    }
   }
 
   @override
   void drawLine(ui.Offset p1, ui.Offset p2, SurfacePaintData paint) {
-    _setUpPaint(paint);
-    _canvasPool.strokeLine(p1, p2);
-    _tearDownPaint();
+    if (_useDomForRendering(paint)) {
+      final SurfacePath path = SurfacePath()
+          ..moveTo(p1.dx, p1.dy)
+          ..lineTo(p2.dx, p2.dy);
+      drawPath(path, paint);
+    } else {
+      _setUpPaint(paint);
+      _canvasPool.strokeLine(p1, p2);
+      _tearDownPaint();
+    }
   }
 
   @override
   void drawPaint(SurfacePaintData paint) {
-    _setUpPaint(paint);
-    _canvasPool.fill();
-    _tearDownPaint();
+    if (_useDomForRendering(paint)) {
+      drawRect(ui.Rect.fromLTWH(0, 0, _bounds.width, _bounds.height), paint);
+    } else {
+      _setUpPaint(paint);
+      _canvasPool.fill();
+      _tearDownPaint();
+    }
   }
 
   @override
   void drawRect(ui.Rect rect, SurfacePaintData paint) {
-    _setUpPaint(paint);
-    _canvasPool.drawRect(rect, paint.style);
-    _tearDownPaint();
+    if (_useDomForRendering(paint)) {
+      html.HtmlElement element = _buildDrawRectElement(rect, paint, 'draw-rect',
+        _canvasPool._currentTransform);
+      _drawElement(element, ui.Offset(math.min(rect.left, rect.right),
+          math.min(rect.top, rect.bottom)), paint);
+    } else {
+      _setUpPaint(paint);
+      _canvasPool.drawRect(rect, paint.style);
+      _tearDownPaint();
+    }
+  }
+
+  /// Inserts a dom element at [offset] creating stack of divs for clipping
+  /// if required.
+  void _drawElement(html.Element element, ui.Offset offset,
+      SurfacePaintData paint) {
+    if (_canvasPool.isClipped) {
+      final List<html.Element> clipElements = _clipContent(
+          _canvasPool._clipStack!,
+          element,
+          ui.Offset.zero,
+          transformWithOffset(_canvasPool._currentTransform, offset));
+      for (html.Element clipElement in clipElements) {
+        rootElement.append(clipElement);
+        _children.add(clipElement);
+      }
+    } else {
+      rootElement.append(element);
+      _children.add(element);
+    }
+    ui.BlendMode? blendMode = paint.blendMode;
+    if (blendMode != null) {
+      element.style.mixBlendMode = _stringForBlendMode(blendMode) ?? '';
+    }
   }
 
   @override
   void drawRRect(ui.RRect rrect, SurfacePaintData paint) {
-    _setUpPaint(paint);
-    _canvasPool.drawRRect(rrect, paint.style);
-    _tearDownPaint();
+    final ui.Rect rect = rrect.outerRect;
+    if (_useDomForRendering(paint)) {
+      html.HtmlElement element = _buildDrawRectElement(rect, paint,
+          'draw-rrect', _canvasPool._currentTransform);
+      _drawElement(element, ui.Offset(math.min(rect.left, rect.right),
+          math.min(rect.top, rect.bottom)), paint);
+      _applyRRectBorderRadius(element.style, rrect);
```

Comment:
Should we call this prior to calling `_drawElement` so that we edit the style while the element is still detached from the live DOM?

(applies to all methods)

Reply:
Done.